volttron/platform/agent/base_historian.py

```diff
@@ -1478,39 +1544,66 @@ def free_count():
-                    if self._record_count < c.rowcount:
-                        self._record_count = 0
+                    if time_tolerance_check and get_error_count_from_db:
+                        # if time_tolerance_check is enabled and this the first time
+                        # we get into this loop, get the count from db
+                        c.execute("SELECT count(ts) from time_error")
+                        error_record_count = c.fetchone()[0]
+                        get_error_count_from_db = False # after this we will reduce count as we delete
+                    if error_record_count > 0:
+                        # if time_error table has records, try deleting those first before outstanding table
+                        _log.info("cache size exceeded limit Deleting data from time_error")
+                        c.execute(
+                            '''DELETE FROM time_error
+                            WHERE ROWID IN
+                            (SELECT ROWID FROM time_error
+                            ORDER BY ROWID ASC LIMIT 100)''')
+                        error_record_count -= c.rowcount
                     else:
-                        self._record_count -= c.rowcount
-                    p = page_count()  #page count doesn't reflect delete without commit
-                    f = free_count() # freelist count does. So using that to break from loop
+                        # error record count is 0, sp set time_error_records to False
+                        self.time_error_records = False
+                        _log.info("cache size exceeded limit Deleting data from outstanding")
+                        c.execute(
+                            '''DELETE FROM outstanding
+                            WHERE ROWID IN
+                            (SELECT ROWID FROM outstanding
+                            ORDER BY ROWID ASC LIMIT 100)''')
+                        if self._record_count < c.rowcount:
+                            self._record_count = 0
+                        else:
+                            self._record_count -= c.rowcount
+                    p = page_count()  # page count doesn't reflect delete without commit
+                    f = free_count()  # freelist count does. So using that to break from loop
                     if f >= min_free_pages:
                         break
                     _log.debug(f" Cleaning cache since we are over the limit. "
                                f"After delete of 100 records from cache"
-                               f" record count is {self._record_count} page count is {p} freelist count is{f}")
+                               f" record count is {self._record_count} time_error record count is {error_record_count} "
+                               f"page count is {p} freelist count is{f}")
 
             except Exception as e:
-                _log.warning(f"Exception when check page count and deleting{e}")
```

Comment:
lets use _log.exception("Unhandled exception happened")  

No as e is necessary and it will give us a nice stack trace.

per https://www.loggly.com/blog/exceptional-logging-of-exceptions-in-python/

Reply:
fixed